File: aoc15/day_17/day_17.ml
let containers = Input.parse_lines int_of_string "input" in

let sum lst = List.fold_left ( + ) 0 lst in
let rec combs lst =
  match lst with
  | [] -> []
  | h :: t ->
      let cs = combs t in
      let res = ([ h ] :: cs) @ List.map (fun c -> h :: c) cs in
      List.filter (fun lst -> sum lst <= 150) res
in
let cs = List.filter (fun lst -> sum lst = 150) (combs containers) in

let first = List.length cs in
let min_len = List.fold_left min 150 (List.map List.length cs) in
let second = List.length (List.filter (fun l -> List.length l = min_len) cs) in
Printf.printf "First: %d\n" first;
Printf.printf "Second: %d\n" second

File: aoc15/go/day_17/day_17.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"

	"gonum.org/v1/gonum/stat/combin"
)

func main() {
	input, _ := os.Open("input")
	scanner := bufio.NewScanner(input)

	containers := []int{}
	for scanner.Scan() {
		c, _ := strconv.Atoi(scanner.Text())
		containers = append(containers, c)
	}

	first := 0
	second := 0
	for n := 0; n < len(containers); n++ {
		if second == 0 && first != 0 {
			second = first
		}
		perms := combin.Combinations(len(containers), n)
		for _, perm := range perms {
			sum := 0
			for _, p := range perm {
				sum += containers[p]
			}
			if sum == 150 {
				first++
			}
		}
	}
	fmt.Println("First:", first)
	fmt.Println("Second:", second)
}
